components/state-component.js

```diff
@@ -0,0 +1,89 @@
+// state-component.js
+
+var _ = require('underscore');
+var jsonpointer = require('jsonpointer');
+var promiseComponent = require('./promise-component');
+
+/**
+ * Combines payloads, from multiple ports, arriving in an asynchronous
+ * fashion. Each port may include an indexBy function, property, or path
+ * that will be used to determine a key that must match payloads from other ports.
+ * If no indexBy is provided then the payload will match any payload from other ports.
+ * A provided onchange function is called when payload is received (from all
+ * required port) with payload having the same key from all other inPorts.
+ *
+ * Usage:
+ *   stateComponent({
+ *     inPorts: {name:{indexBy:function(payload)}} of port names to port options || [] of port names,
+ *     onchange: function(inPayloads, outPayload) {
+ *         inPayloads = {name:payload} of port names to last payload,
+ *         outPayload = previous result of onchange
+ *     }
+ *   });
+ */
+module.exports = function(def){
+    if (!_.isFunction(def.onchange)) throw Error("onchange must be a function");
+    var inPorts = _.isArray(def.inPorts) ? _.object(def.inPorts, []) : def.inPorts;
+    var required = _.compact(_.map(inPorts, function(port, name){
+        return port && port.required && name;
+    }));
+    return promiseComponent(_.defaults({
+        inPorts: _.mapObject(inPorts, function(port, name) {
+            return _.extend({
+                ondata: _.compose(
+                    fireChangeEvent(required, def.onchange),
+                    indexPayload(port && port.indexBy, name)
+                )
+            }, port);
+        })
+    }, def));
+};
+
+function indexPayload(indexBy, name) {
```

Comment:
David has been asking me to include a comment above each function describing what it does, parameters, and return.